server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/webhook/WebhookCallerImplTest.java

```diff
@@ -42,69 +45,128 @@
 
     assertThat(delivery.getHttpStatus()).isEmpty();
     assertThat(delivery.getDurationInMs()).isEmpty();
     // message can be "Failed to connect" or "connect timed out"
     assertThat(delivery.getErrorMessage().get()).contains("connect");
     assertThat(delivery.getAt()).isEqualTo(NOW);
     assertThat(delivery.getWebhook()).isSameAs(webhook);
-    assertThat(delivery.getPayload()).isSameAs(payload);
+    assertThat(delivery.getPayload()).isSameAs(PAYLOAD);
   }
 
   @Test
   public void silently_catch_error_when_url_is_incorrect() throws Exception {
     Webhook webhook = new Webhook(PROJECT_UUID, CE_TASK_UUID, "my-webhook", "this_is_not_an_url");
-    WebhookPayload payload = new WebhookPayload("P1", "{the payload}");
 
-    WebhookDelivery delivery = newSender().call(webhook, payload);
+    WebhookDelivery delivery = newSender().call(webhook, PAYLOAD);
 
     assertThat(delivery.getHttpStatus()).isEmpty();
     assertThat(delivery.getDurationInMs()).isEmpty();
     assertThat(delivery.getError().get()).isInstanceOf(IllegalArgumentException.class);
     assertThat(delivery.getErrorMessage().get()).isEqualTo("unexpected url: this_is_not_an_url");
     assertThat(delivery.getAt()).isEqualTo(NOW);
     assertThat(delivery.getWebhook()).isSameAs(webhook);
-    assertThat(delivery.getPayload()).isSameAs(payload);
+    assertThat(delivery.getPayload()).isSameAs(PAYLOAD);
+  }
+
+  /**
+   * SONAR-8799
+   */
+  @Test
+  public void redirects_should_be_followed_with_POST_method() throws Exception {
+    Webhook webhook = new Webhook(PROJECT_UUID, CE_TASK_UUID, "my-webhook", server.url("/redirect").toString());
+
+    // /redirect redirects to /target
+    server.enqueue(new MockResponse().setResponseCode(307).setHeader("Location", server.url("target")));
+    server.enqueue(new MockResponse().setResponseCode(200));
+
+    WebhookDelivery delivery = newSender().call(webhook, PAYLOAD);
+
+    assertThat(delivery.getHttpStatus().get()).isEqualTo(200);
+    assertThat(delivery.getDurationInMs().get()).isGreaterThanOrEqualTo(0);
+    assertThat(delivery.getError()).isEmpty();
+    assertThat(delivery.getAt()).isEqualTo(NOW);
+    assertThat(delivery.getWebhook()).isSameAs(webhook);
+    assertThat(delivery.getPayload()).isSameAs(PAYLOAD);
+
+    takeAndVerifyPostRequest("/redirect");
+    takeAndVerifyPostRequest("/target");
+  }
+
+  @Test
+  public void redirects_throws_ISE_if_header_Location_is_missing() throws Exception {
+    HttpUrl url = server.url("/redirect");
+    Webhook webhook = new Webhook(PROJECT_UUID, CE_TASK_UUID, "my-webhook", url.toString());
+
+    server.enqueue(new MockResponse().setResponseCode(307));
+
+    WebhookDelivery delivery = newSender().call(webhook, PAYLOAD);
+
+    Throwable error = delivery.getError().get();
+    assertThat(error).isInstanceOf(IllegalStateException.class);
+    assertThat(error).hasMessage("Missing HTTP header 'Location' in redirect of " + url);
```

Comment:
Consider this shorter version: `assertThat(delivery.getError().get()).isInstanceOf(IllegalStateException.class).hasMessage("Missing HTTP header 'Location' in redirect of " + url);`